base/linalg/dense.jl

```diff
@@ -693,6 +698,198 @@ function inv(A::StridedMatrix{T}) where T
 end
 
 """
+    cos(A::Matrix)
+
+Compute the matrix cosine of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is
+used to compute the cosine. Otherwise, the cosine is determined by calling
+`exp`.
+
+# Examples
+```jldoctest
+julia> cos(ones(2, 2))
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function cos(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full(cos(Symmetric(A)))
+    end
+    return real(exp(im*A))
+end
+function cos(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full(cos(Hermitian(A)))
+    end
+    return 0.5 * (exp(im*A) + exp(-im*A))
```

Comment:
Instead of multiplying with (`Float64`) `0.5` here, dived by (`Int`) `2` instead. Otherwise this will promote e.g. `Float32` arrays to `Float64`.